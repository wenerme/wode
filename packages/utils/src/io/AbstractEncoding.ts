/**
 * AbstractEncoding
 *
 * @see https://github.com/mafintosh/abstract-encoding
 */
export interface AbstractEncoding<T> {
  encode(data: T, buffer?: ArrayBuffer, offset?: number): any;

  decode(buffer: ArrayBuffer, start?: number, end?: number): T;
}
