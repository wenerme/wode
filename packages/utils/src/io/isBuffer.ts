// https://github.com/feross/is-buffer/blob/master/index.js
export function isBuffer(obj: any): obj is Buffer {
  return (
    obj != null &&
    obj.constructor != null &&
    typeof obj.constructor.isBuffer === 'function' &&
    obj.constructor.isBuffer(obj)
  );
}
