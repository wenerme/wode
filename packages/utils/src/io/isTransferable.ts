import { getGlobalThis } from '../isomorphics/getGlobalThis';

const globalThis = getGlobalThis();

/**
 * transferable object pass between workers, can work with structuredClone
 *
 * - Chrome 87, FF 103, Safari X, NodeJS X
 *
 * @see {@link https://developer.mozilla.org/en-US/docs/Glossary/Transferable_objects | Transferable objects}
 */
export function isTransferable(v: any): v is TransferableObject {
  _ctors ||= ctors();
  return _ctors.some((ctor) => v instanceof ctor);
}

let _ctors: any[];

function ctors() {
  const o: any = globalThis.window || globalThis || global;
  return [
    o.ArrayBuffer,
    o.MessagePort,
    o.ReadableStream,
    o.WritableStream,
    o.TransformStream,
    o.AudioData,
    o.ImageBitmap,
    o.VideoFrame,
    o.OffscreenCanvas,
    o.RTCDataChannel,
  ].filter(Boolean);
}

export type TransferableObject =
  | Transferable
  | ArrayBuffer
  | MessagePort
  | ReadableStream
  | WritableStream
  | TransformStream
  | AudioData
  | ImageBitmap
  | VideoFrame
  | OffscreenCanvas
  | RTCDataChannel;

declare global {
  interface OffscreenCanvas {}

  interface VideoFrame {}

  interface AudioData {}
}
