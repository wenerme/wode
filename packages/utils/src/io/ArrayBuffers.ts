import { classOf } from '../langs/classOf';
import { isBuffer } from './isBuffer';

/**
 * Various utils to work with {@link ArrayBuffer}
 */
export type ArrayBuffers = {
  /**
   * isArrayBuffer check if the given value is an {@link ArrayBuffer}
   */
  isArrayBuffer(v: any): v is ArrayBuffer;

  /**
   * slice the given view with the given offset and length, will handle the {@link Buffer} as well
   *
   * @see {@link https://nodejs.org/api/buffer.html#bufslicestart-end Buffer.slice}
   */
  slice<T extends ArrayBufferView>(o: T, start?: number, end?: number): T;

  /**
   * asView convert the given value to given {@link TypedArray} view
   *
   * TypedArray can be {@link Buffer}, will avoid copy
   */
  asView<V, C extends ArrayBufferViewConstructor<V>>(
    TypedArray: C,
    v: BufferSource,
    byteOffset?: number,
    byteLength?: number,
  ): V;

  /**
   * toString convert the given {@link BufferSource} to string
   */
  toString(v: BufferSource | string, encoding?: ToStringEncoding): string;
  /**
   * Returns true if encoding is the name of a supported character encoding, or false otherwise.
   */
  isEncoding(v?: string): v is ToStringEncoding;

  toJSON<T = any>(v: BufferSource | string, reviver?: (this: any, key: string, value: any) => any): T;

  /**
   * from convert the given value to {@link ArrayBuffer}
   */
  from(v: string | BufferSource, encoding?: ToStringEncoding): ArrayBuffer;

  /**
   * concat the given {@link BufferSource} to a new {@link ArrayBuffer}
   */
  concat(buffers: Array<BufferSource>, result?: ArrayBuffer, offset?: number): ArrayBuffer;
};

type ToStringEncoding =
  | 'ascii'
  | 'utf16le'
  // | 'utf-16le'
  | 'ucs2'
  | 'ucs-2'
  | 'base64'
  | 'base64url'
  | 'latin1'
  | 'binary'
  | 'utf8'
  | 'utf-8'
  | 'hex';

export const ArrayBuffers = {
  _allowedBuffer: true,
  isArrayBuffer: (v: any): v is ArrayBuffer => {
    return v instanceof ArrayBuffer;
  },
  slice: (o: TypedArray, start?: number, end?: number) => {
    // NodeJS Buffer slice is not the same as UInt8Array slice
    // https://nodejs.org/api/buffer.html#bufslicestart-end
    if (isBuffer(o)) {
      return Uint8Array.prototype.slice.call(o, start, end);
    }
    return o.slice(start, end);
  },
  asView: <V, C extends ArrayBufferViewConstructor<V>>(
    TypedArray: C,
    v: BufferSource,
    byteOffset: number = 0,
    byteLength?: number,
  ): V => {
    if (v instanceof TypedArray && (byteOffset ?? 0) === 0 && byteLength === undefined) {
      return v;
    }
    if (ArrayBuffer.isView(v) || isBuffer(v)) {
      return new TypedArray(v.buffer, v.byteOffset + byteOffset, byteLength ?? v.byteLength);
    }
    return new TypedArray(v, byteOffset, byteLength);
  },
  toString: (buf: BufferSource | string, encoding: ToStringEncoding = 'utf8') => {
    // 'ascii'  'utf16le' | 'ucs2' | 'ucs-2' | 'base64' | 'base64url' | 'latin1' | 'binary' | 'hex'
    if (typeof buf === 'string') {
      return buf;
    }
    if (typeof Buffer !== undefined && ArrayBuffers._allowedBuffer) {
      return Buffer.from(ArrayBuffers.asView(Uint8Array, buf)).toString(encoding);
    }
    // reference
    // https://github.com/feross/buffer/blob/master/index.js
    switch (encoding) {
      case 'hex': {
        let view: Uint8Array = ArrayBuffers.asView(Uint8Array, buf);
        return [...view].map((b) => hexLookupTable[b]).join('');
      }
      case 'base64': {
        let view: Uint8Array = ArrayBuffers.asView(Uint8Array, buf);
        return btoa(String.fromCharCode(...view));
      }
      case 'utf8':
      case 'utf-8':
        return new TextDecoder().decode(buf as any);
      case 'ascii': {
        let view: Uint8Array = ArrayBuffers.asView(Uint8Array, buf);
        return String.fromCharCode(...view.map((v) => v & 0x7f));
      }
      case 'latin1':
      case 'binary': {
        let view: Uint8Array = ArrayBuffers.asView(Uint8Array, buf);
        return String.fromCharCode(...view);
      }
      case 'ucs2':
      case 'ucs-2':
      // case 'utf-16le':
      case 'utf16le': {
        let view: Uint8Array = ArrayBuffers.asView(Uint8Array, buf);
        let res = '';
        // If length is odd, the last 8 bits must be ignored (same as node.js)
        for (let i = 0; i < view.length - 1; i += 2) {
          res += String.fromCharCode(view[i] + view[i + 1] * 256);
        }
        return res;
      }
      default:
        throw new Error(`[ArrayBuffers.toString] Unknown encoding: ${encoding}`);
    }
  },
  toJSON: (v: BufferSource | string, reviver?: (this: any, key: string, value: any) => any) => {
    return JSON.parse(ArrayBuffers.toString(v), reviver);
  },
  alloc: (size: number, fill?: string | number, encoding?: ToStringEncoding) => {
    if (fill !== undefined) {
      if (typeof fill === 'number') {
        return new Uint8Array(size).fill(fill);
      }
      return (ArrayBuffers.asView(Uint8Array, ArrayBuffers.from(fill, encoding)) as Uint8Array).slice(0, size);
    }
    return new ArrayBuffer(size);
  },
  from: (v: string | BufferSource | Array<number>, encoding: ToStringEncoding = 'utf8'): BufferSource => {
    if (!v) {
      return new ArrayBuffer(0);
    }
    if (typeof v === 'string') {
      if (typeof Buffer !== undefined && ArrayBuffers._allowedBuffer) {
        return Buffer.from(v, encoding);
      }

      switch (encoding) {
        case 'utf-8':
        case 'utf8':
          return new TextEncoder().encode(v as string).buffer;
        case 'base64':
          return Uint8Array.from(atob(v.replaceAll(/[^0-9a-zA-Z=+/_ \r\n]/g, '')), (c) => c.charCodeAt(0));
        default:
          throw new Error(`[ArrayBuffers.from] Unknown encoding: ${encoding}`);
      }
    }
    if (v instanceof ArrayBuffer) {
      return v;
    }
    // lost length
    if (ArrayBuffer.isView(v) || isBuffer(v)) {
      if (v.byteOffset !== 0) {
        // return v.buffer.slice(v.byteOffset, v.byteOffset + v.byteLength)
        throw new Error('ArrayBuffers.from do not support view with offset');
      }
      return v.buffer;
    }
    if (Array.isArray(v)) {
      return new Uint8Array(v);
    }
    let type = classOf(v);
    throw new TypeError(`ArrayBuffers.from unsupported type ${type}`);
  },
  isEncoding: (encoding?: string) => {
    switch (String(encoding).toLowerCase()) {
      case 'hex':
      case 'utf8':
      case 'utf-8':
      case 'ascii':
      case 'latin1':
      case 'binary':
      case 'base64':
      case 'ucs2':
      case 'ucs-2':
      case 'utf16le':
        // case 'utf-16le':
        return true;
      default:
        return false;
    }
  },
  concat: (buffers: Array<BufferSource>, result?: ArrayBuffer, offset = 0) => {
    // https://stackoverflow.com/questions/10786128/appending-arraybuffers

    const length = buffers.reduce((a, b) => a + b.byteLength, 0);
    const r = result ? new Uint8Array(result) : new Uint8Array(length);
    for (const buffer of buffers) {
      if (!buffer || !buffer.byteLength) continue;
      let n: Uint8Array;
      if (buffer instanceof ArrayBuffer) {
        n = new Uint8Array(buffer);
      } else if (ArrayBuffer.isView(buffer)) {
        n = new Uint8Array(buffer.buffer, buffer.byteOffset, buffer.byteLength);
      } else {
        throw new Error(`ArrayBuffers.concat unsupported type ${classOf(buffer)}`);
      }
      r.set(n, offset);
      offset += buffer.byteLength;
    }
    return r.buffer;
  },
};

export type TypedArray =
  | Uint8Array
  | Uint8ClampedArray
  | Uint16Array
  | Uint32Array
  | Int8Array
  | Int16Array
  | Int32Array
  | BigUint64Array
  | BigInt64Array
  | Float32Array
  | Float64Array;

type ArrayBufferViewConstructor<T> = new (buffer: ArrayBufferLike, byteOffset?: number, byteLength?: number) => T;

const hexLookupTable = (function () {
  const alphabet = '0123456789abcdef';
  const table = new Array(256);
  for (let i = 0; i < 16; ++i) {
    const i16 = i * 16;
    for (let j = 0; j < 16; ++j) {
      table[i16 + j] = alphabet[i] + alphabet[j];
    }
  }
  return table;
})();
