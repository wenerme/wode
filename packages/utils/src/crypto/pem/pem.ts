import { ArrayBuffers } from '../../io/ArrayBuffers';

export interface Block {
  type: string | 'RSA PRIVATE KEY'; // https://github.com/openssl/openssl/blob/master/include/openssl/pem.h#L35-L60
  header: Record<string, string>;
  bytes: BufferSource;
}

export class PEM {
  static decode(data: string): { block: Block; tail: string; head: string } {
    const match = data.match(
      /^-----BEGIN (?<type>[^\r\n-]+)-----$\r?\n(?<headers>(^[^:\r\n]+:[^\n\r]+\r?\n)+\r?\n)?(?<data>[a-zA-Z0-9/_=\n\r+]+?)^-----END \1-----$\r?\n?/ms,
    );
    if (!match?.groups) throw new Error('Invalid PEM data');
    const { type, headers = '', data: b64 } = match.groups;
    const header = headers
      .split('\n')
      .filter((v) => v.trim())
      .map((h) => {
        const [k, ...v] = h.split(':');
        return [k.trim(), v.join(':').trim()];
      })
      .reduce((a, [k, v]) => ({ ...a, [k]: v }), {});
    return {
      block: {
        type,
        header,
        bytes: ArrayBuffers.from(b64.replaceAll(/[\r\n]/g, ''), 'base64'),
      },
      head: data.slice(0, match.index || 0),
      tail: data.slice((match.index || 0) + match[0].length),
    };
  }
  static encode(block: { type: string; bytes: string | BufferSource; header?: Record<string, string> }): string {
    const { type, header, bytes } = block;
    const headers = Object.entries(header || {})
      .map(([k, v]) => `${k}: ${v}`)
      .join('\n');
    return `-----BEGIN ${type}-----\n${headers}${headers.length ? '\n\n' : ''}${
      ArrayBuffers.toString(bytes, 'base64')
        .match(/.{1,64}/g)
        ?.join('\n') || ''
    }\n-----END ${type}-----\n`;
  }
}
