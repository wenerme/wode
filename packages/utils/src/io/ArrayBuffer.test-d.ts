import { expectTypeOf, test } from 'vitest';
import { ArrayBuffers } from './ArrayBuffers';

test('my types work properly', () => {
  expectTypeOf(ArrayBuffers.asView(Buffer, new Uint8Array())).toMatchTypeOf<Buffer>();
});
