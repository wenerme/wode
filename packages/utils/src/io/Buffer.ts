import { ArrayBuffers } from './ArrayBuffers';
import { isBuffer } from './isBuffer';

/**
 * Buffer is a polyfill version of NodeJS Buffer
 */
export class Buffer extends Uint8Array {
  // constructor(buffer: ArrayBufferLike, byteOffset?: number, length?: number) {
  //   super(buffer, byteOffset, length);
  // }

  static get isPolyfill() {
    return true;
  }

  static isBuffer(v: any): v is Buffer {
    return v instanceof Buffer || isBuffer(v);
  }

  static from(array: string | BufferSource | ArrayLike<number> | Iterable<number>, arg2?: any): Buffer {
    // todo mapfn
    return new Buffer(ArrayBuffers.from(array, arg2) as ArrayBuffer);
  }

  static isEncoding = ArrayBuffers.isEncoding;

  toString(encoding?: string): string {
    return ArrayBuffers.toString(this, encoding as any);
  }
}
