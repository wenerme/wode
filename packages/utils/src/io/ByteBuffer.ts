import { isDefined } from '../langs/isDefined';
import { ArrayBuffers } from './ArrayBuffers';

type AnyBuffer = BufferSource | ArrayBufferLike;

function asBuffer(o: AnyBuffer) {
  if (o instanceof ArrayBuffer) {
    return o;
  }
  if (ArrayBuffer.isView(o)) {
    // 保留 offset&length
    if (o.byteLength !== o.buffer.byteLength) {
      // ArrayBuffer 没有 subarray
      // if ('subarray' in o.buffer) {
      //   return (o.buffer as any).subarray(o.byteOffset, o.byteOffset + o.byteLength);
      // }
      return o.buffer.slice(o.byteOffset, o.byteOffset + o.byteLength);
    }
    return o.buffer;
  }
  return o;
}

// function asView(o: AnyBuffer) {
//   if (o instanceof DataView) {
//     return o;
//   }
//   if (ArrayBuffer.isView(o)) {
//     // 不 clone 也能保留 offset&length
//     return new DataView(o.buffer, o.byteOffset, o.byteLength);
//   }
//   return new DataView(o);
// }

/**
 * @see https://www.egret.uk/docs/egretengine/engine/egret.ByteArray
 * @see https://netty.io/4.1/api/io/netty/buffer/ByteBuf.html
 */
export class ByteBuffer {
  position = 0;

  #buffer: ArrayBufferLike;
  #view: DataView;

  constructor(buffer: AnyBuffer = new ArrayBuffer(0, { maxByteLength: 1024 })) {
    this.#buffer = asBuffer(buffer);
    // note prefer view over buffer, avoid the slice overhead ?
    this.#view = new DataView(this.#buffer);
  }

  get view() {
    return this.#view;
  }

  get buffer(): ArrayBufferLike {
    return this.#buffer;
  }

  set buffer(buffer: AnyBuffer) {
    this.#buffer = asBuffer(buffer);
    this.#view = new DataView(this.#buffer);
  }

  get length() {
    return this.view.byteLength;
  }

  set length(length: number) {
    this.resize(length);
  }

  private willWrite(length: number) {
    if (this.remaining() < length) {
      this.resize(this.length + length);
    }
  }

  resize(newLength: number) {
    // Chrome 111, Nodejs 20
    let buf = this.buffer;
    if (buf.resize && (!isDefined(buf.resizable) || buf.resizable)) {
      buf.resize(newLength);
    } else {
      let newBuffer = new ArrayBuffer(newLength);
      new Uint8Array(newBuffer).set(new Uint8Array(buf));
      this.buffer = newBuffer;
    }
  }

  writeByte(value: number) {
    this.willWrite(1);
    this.view.setUint8(this.position++, value);
  }

  writeBytes(bytes: ArrayBufferLike) {
    this.willWrite(bytes.byteLength);
    new Uint8Array(this.buffer).set(new Uint8Array(bytes), this.position);
    this.position += bytes.byteLength;
  }

  writeInt8(value: number) {
    this.willWrite(1);
    this.view.setInt8(this.position, value);
    this.position += 1;
  }

  writeUint8(value: number) {
    this.willWrite(1);
    this.view.setUint8(this.position, value);
    this.position += 1;
  }

  writeInt16(value: number) {
    this.willWrite(2);
    this.view.setInt16(this.position, value);
    this.position += 2;
  }

  writeUint16(value: number) {
    this.willWrite(2);
    this.view.setUint16(this.position, value);
    this.position += 2;
  }

  writeInt32(value: number) {
    this.willWrite(4);
    this.view.setInt32(this.position, value);
    this.position += 4;
  }

  writeUint32(value: number) {
    this.willWrite(4);
    this.view.setUint32(this.position, value);
    this.position += 4;
  }

  writeInt64(value: bigint) {
    this.willWrite(8);
    this.view.setBigInt64(this.position, value);
    this.position += 8;
  }

  writeUint64(value: bigint) {
    this.willWrite(8);
    this.view.setBigUint64(this.position, value);
    this.position += 8;
  }

  writeFloat32(value: number) {
    this.willWrite(4);
    this.view.setFloat32(this.position, value);
    this.position += 4;
  }

  writeFloat64(value: number) {
    this.willWrite(8);
    this.view.setFloat64(this.position, value);
    this.position += 8;
  }

  writeBoolean(value: boolean) {
    this.writeByte(value ? 1 : 0);
  }

  writeString(value: string): void;
  writeString(value: string, len?: number): void;
  writeString(value: string, len?: number) {
    let encoder = new TextEncoder();
    let bytes = encoder.encode(value);
    this.writeBytes(ArrayBuffers.resize(bytes.buffer, len));
  }

  readByte() {
    return this.view.getUint8(this.position++);
  }

  readBytes(length: number) {
    let bytes = this.buffer.slice(this.position, this.position + length);
    this.position += length;
    return bytes;
  }

  readInt8() {
    let value = this.view.getInt8(this.position);
    this.position += 1;
    return value;
  }

  readUint8() {
    let value = this.view.getUint8(this.position);
    this.position += 1;
    return value;
  }

  readInt16() {
    let value = this.view.getInt16(this.position);
    this.position += 2;
    return value;
  }

  readUint16() {
    let value = this.view.getUint16(this.position);
    this.position += 2;
    return value;
  }

  readInt32() {
    let value = this.view.getInt32(this.position);
    this.position += 4;
    return value;
  }

  readUint32() {
    let value = this.view.getUint32(this.position);
    this.position += 4;
    return value;
  }

  readInt64() {
    let value = this.view.getBigInt64(this.position);
    this.position += 8;
    return safeNumber(value);
  }

  readUint64() {
    let value = this.view.getBigUint64(this.position);
    this.position += 8;
    return safeNumber(value);
  }

  readFloat32() {
    let value = this.view.getFloat32(this.position);
    this.position += 4;
    return value;
  }

  readFloat64() {
    let value = this.view.getFloat64(this.position);
    this.position += 8;
    return value;
  }

  readBoolean() {
    return this.readByte() === 1;
  }

  readString(length: number) {
    let bytes = this.readBytes(length);
    let decoder = new TextDecoder();
    let a = new Uint8Array(bytes);
    let idx = a.indexOf(0);
    if (idx !== -1) {
      bytes = bytes.slice(0, idx);
    }
    return decoder.decode(bytes);
  }

  readHexString(length: number) {
    let bytes = this.readBytes(length);
    return ArrayBuffers.toHex(bytes);
  }

  writeInt24(value: number) {
    this.writeUint8(value & 0xff);
    this.writeUint16(value >> 8);
  }

  readInt24() {
    return this.readUint8() | (this.readUint16() << 8);
  }

  writeZero(length: number) {
    this.writeBytes(new Uint8Array(length).buffer);
  }

  writeValue(typ: TypedValue['type'], val: TypedValue['value']): void;
  writeValue(tv: TypedValue): void;
  writeValue(a: any, b?: any) {
    const tv: TypedValue = typeof a === 'string' ? { type: a, value: b } : a;
    const { type, value, length } = tv;
    switch (type) {
      case 'uint8':
        this.writeUint8(value);
        break;
      case 'byte':
      case 'int8':
        this.writeInt8(value);
        break;
      case 'uint16':
        this.writeUint16(value);
        break;
      case 'int16':
        this.writeInt16(value);
        break;
      case 'uint32':
        this.writeUint32(value);
        break;
      case 'int32':
        this.writeInt32(value);
        break;
      case 'float32':
        this.writeFloat32(value);
        break;
      case 'float64':
        this.writeFloat64(value);
        break;
      case 'string':
        this.writeString(value, length);
        break;
      case 'boolean':
        this.writeBoolean(value);
        break;
      case 'bytes':
        this.writeBytes(value);
        break;
      default:
        throw new Error(`Unknown type ${type}`);
    }
  }

  readInt() {
    return this.readInt32();
  }

  readFloat() {
    return this.readFloat32();
  }

  readDouble() {
    return this.readFloat64();
  }

  remaining() {
    return this.view.byteLength - this.position;
  }

  toUint8Array() {
    return new Uint8Array(this.buffer);
  }

  toHex() {
    return ArrayBuffers.toHex(this.buffer);
  }
}

export interface TypedValue {
  type:
    | 'byte'
    | 'bytes'
    | 'uint8'
    | 'int8'
    | 'uint16'
    | 'int16'
    | 'uint32'
    | 'int32'
    | 'float32'
    | 'float64'
    | 'string'
    | 'boolean';
  value: any;
  length?: number;
}

export function toDump(data: BufferSource) {
  return Array.from(ArrayBuffers.toHex(data).match(/.{1,32}/g) ?? [])
    .map((v, i) => {
      let idxPrefix = `${i.toString(16).padStart(8, '0')}:`;
      let content = Array.from(v.match(/.{4}/g) ?? []).join(' ');
      let chars = Array.from(new Uint8Array(ArrayBuffers.fromHex(v).buffer)).map((v) => {
        let c = String.fromCharCode(v);
        if (v < 32 || v > 126) {
          c = '.';
        }
        return c;
      });
      return [idxPrefix, content.padEnd(40, ' '), chars.join('')].join(' ');
    })
    .join('\n');
}

/**
 * Chrome Websocket Binary format
 * ```
 * 00000000: 0000 0000 0000 0000 0000 0000 0000 0000  ................
 * ```
 */
export function fromDump(dump: string) {
  return ArrayBuffers.fromHex(
    dump
      .split('\n')
      .map((v) => {
        return v.substring(10, 10 + 41).replaceAll(' ', '');
      })
      .join(''),
  );
}

function safeNumber(n: bigint) {
  if (n > Number.MAX_SAFE_INTEGER) {
    return n;
  }
  return Number(n);
}
