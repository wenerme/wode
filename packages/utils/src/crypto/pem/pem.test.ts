import { expect, test } from 'vitest';
import { ArrayBuffers } from '../../io/ArrayBuffers';
import { PEM } from './pem';

const { decode, encode } = PEM;

test('pem', () => {
  const pem = `-----BEGIN RSA PRIVATE KEY-----
Proc-Type: 4,ENCRYPTED
DEK-Info: DES-EDE3-CBC,ABC

MDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAw
MDAwMDAwMDAwMDAwMDAwMA==
-----END RSA PRIVATE KEY-----
`;
  const input = `Nice\n${pem}Hello
`;
  const out = {
    block: {
      type: 'RSA PRIVATE KEY',
      header: {
        'Proc-Type': '4,ENCRYPTED',
        'DEK-Info': 'DES-EDE3-CBC,ABC',
      },
      bytes: ArrayBuffers.from('0'.repeat(64)),
    },
    head: 'Nice\n',
    tail: 'Hello\n',
  };
  expect(decode(input)).toEqual(out);
  expect(encode(out.block)).toBe(pem);
});
test('cases', () => {
  for (const b of [
    {
      type: 'RSA PRIVATE KEY',
      bytes: 'Hello',
    },
    {
      type: 'RSA PRIVATE KEY',
      header: {
        Name: 'N-:1/=',
        Age: '1234',
      },
      bytes: ArrayBuffers.from('Nice'),
    },
  ]) {
    const s = encode(b);
    const act = decode(s);
    expect({
      ...act.block,
      bytes: ArrayBuffers.toString(act.block.bytes),
    }).toStrictEqual({
      header: {},
      ...b,
      bytes: ArrayBuffers.toString(b.bytes),
    });
    expect(s).matchSnapshot();
  }
});
