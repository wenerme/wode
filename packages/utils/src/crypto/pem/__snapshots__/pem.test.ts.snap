// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cases 1`] = `
"-----BEGIN RSA PRIVATE KEY-----
SGVsbG8=
-----END RSA PRIVATE KEY-----
"
`;

exports[`cases 2`] = `
"-----BEGIN RSA PRIVATE KEY-----
Name: N-:1/=
Age: 1234

TmljZQ==
-----END RSA PRIVATE KEY-----
"
`;
